README.md
# Flicker

Flicker is a multi-disciplinary art project incorporating software engineering practices into the making of performance art.

## Concept

* Live recorded video is cached and projected onto a wall, creating a juxtaposition of current and past events.
* Audio and visual effects are synced to an computer controlled metronome.
* A "Flicker" is a recording of a past event played in the present time.

## Prerequisites

### Hardware
* MacBook Pro, Early 2015
* Logitech C922x webcam
* Optoma EH416 projector via VGA cable

### Software
* [Ableton Live 9](https://www.ableton.com/en/)
* [openFrameworks 0.10.1](https://openframeworks.cc/versions/v0.10.1/of_v0.10.1_osx_release.zip)
* The source code in this repository
* Other openFrameworks libraries
   * [ofxAbletonLink](https://github.com/2bbb/ofxAbletonLink)
   * [ofxAbletonLive](https://github.com/genekogan/ofxAbletonLive)
   * [ofxBenG](https://github.com/bgoosman/ofxBenG)
   * [ofxBlackmagic](https://github.com/kylemcdonald/ofxBlackmagic)
   * [ofxEasing](https://github.com/arturoc/ofxEasing)
   * [ofxIntersection](https://github.com/robandrose/ofxIntersection)
   * [ofxKinect](https://github.com/ofTheo/ofxKinect)
   * [ofxMaxim](https://github.com/falcon4ever/ofxMaxim)
   * [ofxMidi](https://github.com/danomatika/ofxMidi)
   * [ofxMidiFighterTwister](https://github.com/armadillu/ofxMidiFighterTwister)
   * [ofxOpenCv](https://openframeworks.cc/documentation/ofxOpenCv/)
   * [ofxOsc](https://openframeworks.cc/documentation/ofxOsc/)
   * [ofxPS3EyeGrabber](https://github.com/bakercp/ofxPS3EyeGrabber)
   * [ofxPlaymodes](https://github.com/bgoosman/ofxPlaymodes)
   * ofxPoco (comes with ofx 0.10.1)
   * [ofxSyphon](https://github.com/astellato/ofxSyphon)

## Setup

1. Connect the webcam and projector to the laptop
2. Point webcam at desired action
3. Point projector on a surface
4. Compile and run Flicker code
5. System should auto detect the hardware and project a live video feed onto the chosen surface


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- Nothing at the moment

## [2.0.0] - 2019-4-1
### Added
- Light control via OSC messages sent over ethernet to an ETC Element
- Test button plays audio and massages the lights for peace of mind
- Media
   - [performance]()
   - [rehearsal](https://www.dropbox.com/sh/iexaglcbw07zlbj/AAAzBb6Eqb3XLuQ9mN39sQ2Ra?dl=0)

## [1.5.0] - 2019-2-17
### Added
- venue: 12 Minutes Max audition at Base
- setup: ladder, chair
- Media
   - [performance](https://vimeo.com/317951665)

## [1.4.0] - 2019-1-21
### Added
- Camera previews on app window
- Number of Ableton Link peers on app window
- Names of connected monitors on app window
- Go button starts the performance, and a colored background indicates running
- Set levels on ETC Element lightboard via MIDI
- Execute a function for all linearly interpolated values of a range in a set number of beats
- Media
   - [performance](https://www.dropbox.com/s/tniv8zyec0jc3de/IMG_2768.mov?dl=0)
   - [rehearsal](https://www.dropbox.com/sh/ykcqou1ojxk7l1m/AABV4-qVop_dipSbKFXg7Obqa?dl=0)
   - [images](https://www.dropbox.com/sh/84pf8ghbpi7nld4/AACKp6Tk1vpWRHz0mNgy3Cj8a?dl=0)

## [1.3.0] - 2019-01-07
### Added
- solid_color_view draws a square with solid background color on a window
- window_manager::getWindowForMonitor(monitorName) selects fetches the window for a present monitor
- monitor_manager::excludeMonitor(monitorName) will prevent monitor from being added
- stream_manager::excludeDevice(deviceName) will prevent device from being added
- window::getMonitor()
- Defined Optoma 1080p as default monitor for video
- Defined Optoma WXGA as default monitor for light
- During recording phase, video projector is faded black while light projector is unfaded using the solid_color_view
- Flicker length can vary over time, creating a rhythm of white/black and gas clicks.
- Media
   - [performance](https://www.dropbox.com/sh/kavy5i113q0f0qa/AABDZlJ9aIg6h-DCGEtqdB5Ia?dl=0)
   - [rehearsal](https://www.dropbox.com/sh/ro03mfzacmmjoeu/AACyHA4E6OudOhF54bC11-Mha?dl=0)
   - [images](https://www.dropbox.com/sh/h96nfvk4u1yw9ih/AAAg8KZ4s6Piz8TEfTFe737Za?dl=0)

## [1.2.0] - 2018-12-18
### Added
- venue: Stone Way Wide Open Mic
- setup: projector, laptop, and webcam become optional
- [5 minute portable audio mix](https://www.dropbox.com/s/6963o05o7lv1tvq/Flicker.wav?dl=0)

## [1.1.0] - 2018-12-17
### Added
- Ableton: drier egg timer, printer, tick
- setup: optionally install all hardware ahead of time
- video
   - [performance](https://vimeo.com/307926111)
   - [improvised material](https://www.dropbox.com/sh/3ra3wx4kre13q9r/AAA2LOrc37Lagq3tcwH42_03a?dl=0)

### Fixed
- setup: ensure audience can hear audio 

## [1.0.0] - 2018-12-03
### Added
- venue: SH*T GOLD
- Ableton: hammer on metal, gas click 1 & 2, volca loop 6
- [movement vocabulary](https://www.dropbox.com/sh/tz0bht4j0wkbmnb/AADISMIJ_5A9b47Rmvd1KMtua?dl=0)
- setup: install hardware right before performance starts
- video
   - [performance](https://vimeo.com/307925301)
   - [rehearsal](https://www.dropbox.com/sh/qvt56qeqvs92wb1/AAAcmIStd5B1zFth13NdRLura?dl=0)
   - [improvised material](https://www.dropbox.com/sh/wamgz512igkkch6/AABcvAux8IHUHlR5lQB-Ld8fa?dl=0)


src/ofApp.cpp
#include "ofApp.h"

ofApp::ofApp(std::shared_ptr<ofAppBaseWindow> myWindow) :
myWindow(myWindow) {
    ofSetVerticalSync(true);
}

ofApp::~ofApp() {
    ofxBenG::ableton()->stopAll();
    delete streamManager;
}

//--------------------------------------------------------------
void ofApp::setup() {
    ofSetLogLevel(OF_LOG_WARNING);
//    int const desiredWidth = 720;
//    int const desiredHeight = 486;
//    float const desiredFps = 59.94;
//    int const desiredWidth = 1920;
//    int const desiredHeight = 1080;
//    float const desiredFps = 59.94;
    int const desiredWidth = 1280;
    int const desiredHeight = 720;
    float const desiredFps = 60;
    int const defaultBufferSize = 60;
    double const beatsPerMinute = 20;
    streamManager = new ofxBenG::stream_manager(desiredWidth, desiredHeight, desiredFps, defaultBufferSize);
    streamManager->excludeDevice("FaceTime HD Camera");
    ofAddListener(streamManager->onVideoStreamAdded, this, &ofApp::onVideoStreamAdded);
    ofAddListener(streamManager->onVideoStreamRemoved, this, &ofApp::onVideoStreamRemoved);
    monitorManager = new ofxBenG::monitor_manager();
    monitorManager->excludeMonitor("HP Pavilion32");
    monitorManager->excludeMonitor("Color LCD"); 
    monitorManager->excludeMonitor("BenQ GW2765");
    ofAddListener(monitorManager->onMonitorAdded, this, &ofApp::onMonitorAdded);
    ofAddListener(monitorManager->onMonitorRemoved, this, &ofApp::onMonitorRemoved);
    windowManager = new ofxBenG::window_manager();
    windowManager->setFullscreen(true);
    timeline = new ofxBenG::timeline(4);
    lightBoard.setup("169.254.1.42", 3000);
    parameters.setName("settings");
    parameters.add(fadeInTime.set("fadeInTime", 1.0, 0.0, 10.0));
    parameters.add(fadeOutTime.set("fadeOutTime", 1.5, 0.0, 10.0));
    parameters.add(fadeInPreTime.set("fadeInPreTime", 0.5, 0.0, 10.0));
    parameters.add(lightMin.set("lightMin", 40.0, 0.0, 100.0));
    parameters.add(lightMax.set("lightMax", 75.0, 0.0, 100.0));
    parameters.add(aroundTheWorldRunning.set("aroundTheWorldRunning", false));
    gui.setup(parameters);
    gui.add(testButton.setup("test 1", 150, 25));
    gui.add(testButton2.setup("test 2", 150, 25));
    gui.add(blackoutButton.setup("blackout", 150, 25));
    gui.add(goButton.setup("Go", 150, 150));
    goButton.addListener(this, &ofApp::startPerformance);
    testButton.addListener(this, &ofApp::testButtonPressed);
    testButton2.addListener(this, &ofApp::testButton2Pressed);
    blackoutButton.addListener(this, &ofApp::onBlackoutButtonPressed);
    renderer = new ofxPm::BasicVideoRenderer;
    ofxBenG::ableton()->setup(beatsPerMinute, 8.0);
    ofxBenG::ableton()->update();
    ofxBenG::ableton()->stopAll();
}

void ofApp::update() {
    ofxBenG::ableton()->update();
    streamManager->update();
    monitorManager->update();
    timeline->update();
}

void ofApp::draw() {
    if (isRunning) {
        ofBackground(ofColor::darkMagenta);
    }

    std::stringstream strm;
    strm << "fps: " << ofGetFrameRate();
    ofSetWindowTitle(strm.str());

    float topPadding = 25;

    gui.draw();

    float x = gui.getPosition()[0];
    float y = gui.getHeight() + gui.getPosition()[1] + topPadding;

    auto stream = streamManager->getStream(0);
    if (stream != nullptr) {
        ofDrawBitmapString(ofToString(stream->getFps()), ofPoint(x, y));
        y += topPadding;
    }

    ofDrawBitmapString("bpm: " + ofToString(ofxBenG::ableton()->getTempo()), ofPoint(x, y));
    y += topPadding;

    ofDrawBitmapString("beat: " + ofToString(ofxBenG::ableton()->getBeat()), ofPoint(x, y));
    y += topPadding;

    float const previewWidth = 160;
    float const previewHeight = 90;

    for (auto monitor : monitorManager->getMonitors()) {
        ofDrawBitmapString("Monitor: " + monitor->getName(), ofPoint(x, y));
        y += topPadding;
    }

    for (auto headerView : cameraPreviews) {
        headerView->draw(x, y, previewWidth, previewHeight);
        y += previewHeight + 5;
    }
}

void ofApp::audioOut(float* output, int bufferSize, int nChannels) {
    for (int i = 0; i < bufferSize; i++) {
        float mix = ofxBenG::audio::getInstance()->getMix();
        output[nChannels * i] = mix;
        output[nChannels * i + 1] = mix;
    }
}

void ofApp::onVideoStreamAdded(ofxBenG::video_stream &stream) {
    std::cout << "onVideoStreamAdded stream: " << stream.getDeviceName() << std::endl;

    auto headerView = new ofxBenG::header_view(stream.makeHeader(0));
    cameraPreviews.push_back(headerView);

    auto window = windowManager->getWindowWithNoStream();
    if (window != nullptr && window->getMonitorName() != "Optoma WXGA") {
        window->setStream(&stream);
    }
}

void ofApp::onVideoStreamRemoved(ofxBenG::video_stream &stream) {
    std::cout << "onVideoStreamRemoved stream: " << stream.getDeviceName() << std::endl;
}

void ofApp::onMonitorAdded(ofxBenG::monitor &monitor) {
    std::cout << "Monitor added: " << monitor.toString() << std::endl;
    auto window = windowManager->makeWindow(myWindow);
    window->setMonitor(&monitor);
    window->setStream(streamManager->getUnusedStream());
}

void ofApp::onMonitorRemoved(ofxBenG::monitor &monitor) {
    std::cout << "Monitor removed: " << monitor.toString() << std::endl;
}

void ofApp::startPerformance() {
    isRunning = true;
    timeline->scheduleNextWholeBeat(new ofxBenG::generic_action([this]() {
        CUE(stopClip("muffle", "fumbling around"));
        CUE(stopClip("piano treble drone", "piano treble drone 1"));
        CUE(stopClip("guitar", "guitar 2"));
        CUE(stopClip("grand piano", "1"));
        CUE(stopClip("grand piano", "2"));
        CUE(stopClip("grand piano", "3"));
        CUE(stopClip("grand piano", "4"));
        CUE(stopClip("piano bass drone", "piano bass drone 1"));
        CUE(stopClip("guitar drone", "guitar drone uplifting"));
        CUE(stopClip("violin", "violin forward"));
        CUE(ofxBenG::ableton()->setClockToZero());
        CUE(ofxBenG::ableton()->play());
        CUE(ofxBenG::ableton()->getMasterTrack()->getDevice("Auto Filter")->getParameter("Frequency")->setValue(135));
        CUE(ofxBenG::ableton()->getTrack("violin")->getDevice("Simple Delay")->setEnabled(false));
        CUE(ofxBenG::ableton()->getTrack("violin")->getDevice("Ping Pong Delay")->setEnabled(false));
        CUE(ofxBenG::ableton()->getTrack("violin")->getDevice("Ping Pong Delay")->getParameter("Freeze")->setValue(0));
        CUE(ofxBenG::ableton()->getTrack("guitar drones")->setVolume(0));
        CUE(ofxBenG::ableton()->getTrack("guitar drones")->setMute(false));
        CUE(blackout());
        CUE(playClip("muffle", "fumbling around"));
        CUE(playClip("grand piano", "1"));
        CUE_IN_SECONDS(1, fadeLights(60.0, lightMin, lightMax));
        CUE_IN_SECONDS(1, playClip("guitar drones", "guitar drone uplifting"));
        CUE_IN_SECONDS(1, fadeVolume(ofxBenG::ableton()->getTrack("guitar drones"), 60, 0, 0.7));
        CUE_IN_SECONDS(oneMinute / 2, playClip("piano treble drone", "piano treble drone 1"));
        CUE_IN_SECONDS(1 * oneMinute, executeTransition0());
        CUE_IN_SECONDS(2 * oneMinute, executeTransition1());
        CUE_IN_SECONDS(3 * oneMinute, executeTransition2());
        CUE_IN_SECONDS(4 * oneMinute, executeTransition3());
        CUE_IN_SECONDS(showLength, executeTransition4());
    }));
}

void ofApp::testButtonPressed() {
    timeline->cue(fadeLights(5.0, lightMin, lightNormal));
    CUE(playClip("test", "1"));
}

void ofApp::testButton2Pressed() {
    timeline->cue(aroundTheWorld());
    CUE(playClip("steve", "airplane scrub short"));
    CUE_IN_SECONDS(1.0, stopClip("steve", "airplane scrub short"));
    CUE_IN_SECONDS(3.0, playClip("steve", "airplane scrub short"));
    timeline->cue(fadeVolume(ofxBenG::ableton()->getTrack("guitar drones"), 1.0, volumeMin, volumeMax));
}

void ofApp::onBlackoutButtonPressed() {
    timeline->cue(blackout());
}

//--------------------------------------------------------------
void ofApp::keyReleased(int key) {
    if (key == 's') {
        startPerformance();
    }
    
    if (key == 'q') {
        ofExit(0);
    }
}

//--------------------------------------------------------------
void ofApp::keyPressed(int key) {
}

//--------------------------------------------------------------
void ofApp::mouseMoved(int x, int y) {
    
}

//--------------------------------------------------------------
void ofApp::mouseDragged(int x, int y, int button) {
    
}

//--------------------------------------------------------------
void ofApp::mousePressed(int x, int y, int button) {
    
}

//--------------------------------------------------------------
void ofApp::mouseReleased(int x, int y, int button) {
    
}

//--------------------------------------------------------------
void ofApp::mouseEntered(int x, int y) {
    
}

//--------------------------------------------------------------
void ofApp::mouseExited(int x, int y) {
    
}

//--------------------------------------------------------------
void ofApp::windowResized(int w, int h) {
    
}

//--------------------------------------------------------------
void ofApp::gotMessage(ofMessage msg) {
    
}

//--------------------------------------------------------------
void ofApp::dragEvent(ofDragInfo dragInfo) {
    
}


src/ofApp.h
#pragma once

#include "ofMain.h"
#include "beat_action.h"
#include "video_stream.h"
#include "monitor.h"
#include "stream_manager.h"
#include "monitor_manager.h"
#include "window_manager.h"
#include "etc_element.h"
#include "ofxAbletonLive.h"
#include "ofxMidi.h"
#include "ofxGui.h"
#include <functional>
#include <algorithm>

#define TO_SECONDS(x) ((float)(x) / (float)1e6)

#define PROMISE(action) \
new ofxBenG::generic_action([&]() {\
std::cout << #action << std::endl\
action;\
}

class ofApp : public ofBaseApp {
public:
    ofApp(std::shared_ptr<ofAppBaseWindow> mainWindow);
    ~ofApp();
    
    void setup();
    void update();
    void draw();
    void keyPressed(int key);
    void keyReleased(int key);
    void mouseMoved(int x, int y);
    void mouseDragged(int x, int y, int button);
    void mousePressed(int x, int y, int button);
    void mouseReleased(int x, int y, int button);
    void mouseEntered(int x, int y);
    void mouseExited(int x, int y);
    void windowResized(int w, int h);
    void dragEvent(ofDragInfo dragInfo);
    void gotMessage(ofMessage msg);

    void onVideoStreamAdded(ofxBenG::video_stream &stream);
    void onVideoStreamRemoved(ofxBenG::video_stream &stream);
    void onMonitorAdded(ofxBenG::monitor &monitor);
    void onMonitorRemoved(ofxBenG::monitor &monitor);
    void audioOut(float *output, int bufferSize, int nChannels);
    void testButtonPressed();
    void testButton2Pressed();
    void startPerformance();
    void onBlackoutButtonPressed();
    
    void muteAll() {
        ofxBenG::ableton()->getMasterTrack()->getDevice("Auto Filter")->getParameter("Frequency")->setValue(20);
    }
    
    void blackout() {
        for (int i = channelRange[0]; i < channelRange[1]; i++) {
            lightBoard.setChannel(i, 0);
        }
    }
    
    void fadeLights(float durationBeats, int start, int end) {
        for (int i = 0; i < allLightsUsed.size(); i++) {
            timeline->cue(fadeLight(allLightsUsed[i], seconds, start, end));
        }
    }
    
    void fadeLights(std::vector<int> channels, float durationBeats, int start, int end) {
        for (int i = 0; i < channels.size(); i++) {
            timeline->cue(fadeLight(channels[i], seconds, start, end));
        }
    }
    
    void executeTransition0() {
        timeline->cue(playClip("grand piano", "2"));
    }
    
    void executeTransition1() {
        auto frequency = ofxBenG::ableton()->getMasterTrack()->getDevice("Auto Filter")->getParameter("Frequency");
        timeline->cue(fadeParameter(frequency, 1.0, filterMax, filterMin));
        timeline->cue(PROMISE(playClip("guitar", "guitar 2")));
        timeline->cue(PROMISE(playClip("piano bass drone", "piano bass drone 1")));
        timeline->cue(PROMISE(playClip("grand piano", "3")));
        timeline->cueInSeconds(0.5, fadeLights(allLightsUsed, 0.5, lightNormal, lightMin));
        timeline->cueInSeconds(1.5, PROMISE(flickerLight(sideRight[1], 0.3)));
        timeline->cueInSeconds(4, fadeLights(allLightsUsed, 2, lightMin, lightNormal));
        timeline->cueInSeconds(4, fadeParameter(frequency, 2, filterMin, filterMax));
    }
    
    void executeTransition2() {
        auto frequency = ofxBenG::ableton()->getMasterTrack()->getDevice("Auto Filter")->getParameter("Frequency");
        timeline->cue(fadeParameter(frequency, 1.0, filterMax, filterMin));
        timeline->cue(playClip("grand piano", "4"));
        timeline->cueInSeconds(0.5, fadeLights(allLightsUsed, 0.5, lightNormal, lightMin));
        timeline->cueInSeconds(1.5, flickerLight(sideRight[1], 0.3));
        timeline->cueInSeconds(4, fadeLights(allLightsUsed, 2, lightMin, lightNormal));
        timeline->cueInSeconds(4, fadeParameter(frequency, 2, filterMin, filterMax));
    }
    
    void executeTransition3() {
        auto frequency = ofxBenG::ableton()->getMasterTrack()->getDevice("Auto Filter")->getParameter("Frequency");
        timeline->cue(fadeParameter(frequency, 1.0, filterMax, filterMin));
        timeline->cue(playClip("grand piano", "5"));
        timeline->cue(playClip("violin", "violin forward"));
        timeline->cueInSeconds(3.0, new ofxBenG::generic_action([this]() {
            ofxBenG::ableton()->getTrack("violin")->getDevice("Ping Pong Delay")->setEnabled(true);
            ofxBenG::ableton()->getTrack("violin")->getDevice("Ping Pong Delay")->getParameter("Freeze")->setValue(1);
        }));
        timeline->cueInSeconds(0.5, fadeLights(allLightsUsed, 0.5, lightNormal, lightMin));
        timeline->cueInSeconds(1.0, PROMISE(ofxBenG::ableton()->getTrack("guitar drones")->setMute(true)));
        timeline->cueInSeconds(1.5, flickerLight(sideRight[1], 0.3));
        timeline->cueInSeconds(4, fadeLights(allLightsUsed, 2, lightMin, lightNormal));
        timeline->cueInSeconds(4, fadeParameter(frequency, 2, filterMin, filterMax));
        timeline->cueInSeconds(6, fadeLights(allLightsUsedExceptChannelTwo, 42, lightNormal, lightMin));
        timeline->cueInSeconds(48, fadeLight(2, 12, lightNormal, lightMin));
        timeline->cueInSeconds(48, fadeParameter(frequency, 8, filterMax, 0));
    }
    
    void executeTransition4() {
        timeline->cueInSeconds(6, new ofxBenG::generic_action([this]() {
            ofxBenG::ableton()->stopAll();
            timeline->clearScheduledActions();
            timeline->cueInSeconds(4.0, fadeLights(houseLights, 2.0, lightMin, lightNormal));
        }));
    }

    void playSample(ofxMaxiSample *sample) {
        ofxBenG::audio::getInstance()->playSample(sample);
    }

    void playClip(std::string track, std::string clip) {
        ofxBenG::ableton()->playClip(track, clip);
    }

    void playClip(ofxAbletonLiveClip *clip) {
        clip->play();
    }
    
    void stopClip(std::string track, std::string clip) {
        ofxBenG::ableton()->stopClip(track, clip);
    }
    
    void setColor(ofxBenG::single_color_view *view, ofColor color) {
        view->setColor(color);
    }
    
    void setSubmaster(int faderNumber, int percent) {
        lightBoard.setSubmaster(faderNumber, percent);
    }
    
    void setChannel(int faderNumber, int percent) {
        lightBoard.setChannel(faderNumber, percent);
    }
    
    void flickerLight(int channel, float duration) {
        timeline->cue(PROMISE(setChannel(channel, lightMax)));
        timeline->cueInSeconds(duration, PROMISE(setChannel(channel, lightMin)));
    }
    
    void aroundTheWorld() {
        aroundTheWorldRunning = true;
        
        for (int i = 0; i < aroundTheWorldLights.size(); i++) {
            int const channel = aroundTheWorldLights[i];
            timeline->cueInSeconds(i * fadeInTime, fadeLight(channel, fadeInTime, lightMin, lightMax));
            timeline->cueInSeconds(i * fadeInTime + fadeInTime, fadeLight(channel, fadeOutTime, lightMax, lightMin));
        }
       
        timeline->cueInSeconds(fadeInTime * aroundTheWorldLights.size(), new ofxBenG::generic_action([&]() {
            if (aroundTheWorldRunning) timeline->cue(PROMISE(aroundTheWorld()));
        }));
    }

    ofxBenG::lerp_action *fadeLight(int faderNumber, float seconds, int startLevel, int endLevel) {
        return new ofxBenG::lerp_action(seconds, [=](float lerpValue, float min, float max) {
            int const faderLevel = (int)ofMap(lerpValue, min, max, (float)startLevel, (float)endLevel);
            lightBoard.setChannel(faderNumber, faderLevel);
        });
    }
    
    ofxBenG::lerp_action *fadeVolume(ofxAbletonLiveTrack *track, float seconds, float startLevel, float endLevel) {
        return new ofxBenG::lerp_action(seconds, [=](float lerpValue, float min, float max) {
            track->setVolume(ofMap(lerpValue, min, max, startLevel, endLevel));
        });
    }
    
    ofxBenG::lerp_action *fadeParameter(ofxAbletonLiveParameter *parameter, float seconds, float startLevel, float endLevel) {
        return new ofxBenG::lerp_action(seconds, [=](float lerpValue, float min, float max) {
            float newFrequency = ofMap(lerpValue, min, max, startLevel, endLevel);
            parameter->setValue(newFrequency);
        });
    }

    // GUI
    ofxButton testButton;
    ofxButton testButton2;
    ofxButton blackoutButton;
    ofxButton goButton;
    ofParameter<float> fadeOutTime;
    ofParameter<float> fadeInTime;
    ofParameter<float> fadeInPreTime;
    ofParameter<bool> aroundTheWorldRunning;
    ofParameter<float> lightMin;
    ofParameter<float> lightMax;
    ofParameterGroup parameters;
    ofxPanel gui;

    // Lights
    ofxBenG::etc_element_osc_proxy lightBoard;
    std::string lightBoardIp = "169.254.1.42";
    int lightBoardPort = 3000;
    std::vector<int> lights{5, 12, 13, 14, 16, 17, 25};
    ofPoint channelRange = {0, 120};
    std::vector<int> sideRight{11, 12, 13, 14};
    std::vector<int> sideLeft{15, 16, 17, 18};
    std::vector<int> backs{25};
    std::vector<int> fronts{1, 2, 3};
    std::vector<int> allLightsUsed{2, 11, 12, 15, 16, 25};
    std::vector<int> allLightsUsedExceptChannelTwo{11, 12, 15, 16, 25};
    std::vector<int> houseLights{1, 2, 3, 4, 5, 6, 11, 12, 13, 14, 15, 16, 17, 18, 22};
    std::vector<int> aroundTheWorldLights = {25, 12, 5, 16};
    float const lightNormal = 75;

    // Video
    ofxBenG::video_stream *stream = nullptr;
    ofxPm::VideoBuffer *buffer = nullptr;
    ofxPm::VideoHeader *header = nullptr;
    ofxPm::BasicVideoRenderer *renderer = nullptr;
    ofTexture *holdFrame = nullptr;
    float recordingFps;
    
    // Audio
    int const beatsInBar = 4;
    float const volumeOff = 0.0;
    float const volumeMin = 0.5;
    float const volumeMax = 0.75;
    float const filterMin = 60;
    float const filterMax = 135;

    // App
    std::shared_ptr<ofAppBaseWindow> myWindow;
    ofxBenG::stream_manager *streamManager;
    ofxBenG::monitor_manager *monitorManager;
    ofxBenG::window_manager *windowManager;
    ofxBenG::timeline *timeline;
    std::vector<ofxBenG::header_view*> cameraPreviews;
    bool isRunning = false;
    int const oneMinute = 60;
    int const showLength = oneMinute * 5;
};

class blackout_view : public ofxBenG::window_view {
public:
    virtual void draw(ofPoint size) {
        ofPushStyle();
        ofSetColor(ofColor::black);
        ofDrawRectangle(0, 0, size[0], size[1]);
        ofPopStyle();
    }
};

